pkg/controllers/nodeclaim/termination/controller.go

```diff
@@ -83,21 +83,27 @@ func (c *Controller) Finalize(ctx context.Context, nodeClaim *v1beta1.NodeClaim)
 				return reconcile.Result{}, err
 			}
 		}
+		if nodeClaim.Status.ProviderID != "" {
+			requeue, err := nodeclaimutil.TerminationHandler(ctx, c.kubeClient, nodeClaim, node, c.cloudProvider)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			if requeue {
+				return reconcile.Result{Requeue: requeue}, nil
+			}
+		}
 	}
 	// We wait until all the nodes associated with this nodeClaim have completed their deletion before triggering the finalization of the nodeClaim
 	if len(nodes) > 0 {
 		return reconcile.Result{}, nil
 	}
-	if nodeClaim.Status.ProviderID != "" {
-		if err = c.cloudProvider.Delete(ctx, nodeClaim); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
-			// We expect cloudProvider to emit a Retryable Error when the underlying instance is not terminated and if that
-			// happens, we want to re-enqueue reconciliation until we terminate the underlying instance before removing
-			// finalizer from the nodeClaim.
-			if cloudprovider.IsRetryableError(err) {
-				return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
-			}
-			return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+	// If there are no nodes associated with a nodeClaim, and it's status condition was never updated to terminating, we update it here
```

Comment:
Why would a node claim not have the status condition set? In what we have here, we would have an exponential backoff and expect to have the nodeclaim status condition set

Reply:
I understood what @engedaam said in this comment as why we will have nodeClaim with status condition not set and my answer to it is that if there are no nodes found for the nodeClaim in `AllNodesForNodeClaim` then the status condition on it would have never been set and hence this logic here. Unless we don't want to set status condition on nodeClaim without any nodes.
@jonathan-innis not sure if that's what you meant as well?